pkg/kv/store.go

```diff
@@ -0,0 +1,114 @@
+	ErrOperationFailed     = errors.New("operation failed")
+	ErrPredicateFailed     = errors.New("predicate failed")
+	ErrSetupFailed         = errors.New("setup failed")
+	ErrUnknownDriver       = errors.New("unknown driver")
+)
+
+// Driver define the root interface to access the kv database.
+// Each kv provider implements a driver which provides Store instance.
+type Driver interface {
+	// Open access to the database store. Implementation should give access to the same storage based on the dsn.
+	// Implementation can return the same Storage instance based on dsn or new one as long as it provides access to
+	// the same storage.
+	Open(ctx context.Context, dsn string) (Store, error)
+}
+
+type Store interface {
+	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
+	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Set stores the given value, overwriting an existing value if one exists
+	Set(ctx context.Context, key, value []byte) error
+
+	// SetIf returns an ErrPredicateFailed error if the key with valuePredicate passed
+	//  doesn't match the currently stored value. SetIf is a simple compare-and-swap operator:
+	//  valuePredicate is either the existing value, or nil for no previous key exists.
+	//  this is intentionally simplistic: we can model a better abstraction on top, keeping this interface simple for implementors
+	SetIf(ctx context.Context, key, value, valuePredicate []byte) error
+
+	// Delete will delete the key, no error in if key doesn't exist
+	Delete(ctx context.Context, key []byte) error
+
+	// Scan returns entries that can be read by key order, starting at or after the `start` position
+	Scan(ctx context.Context, start []byte) (Entries, error)
+
+	// Close access to the database store. After calling Close the instance is unusable.
+	Close()
+}
+
+// Entries used to enumerate over Scan results
+type Entries interface {
+	// Next should be called first before access Entry.
+	// it will process the next entry and return true if it was successful, and false when none or error.
+	Next() bool
+
+	// Entry current entry read after calling Next, set to nil in case of an error or no more entries.
+	Entry() *Entry
+
+	// Err set to last error by reading or parse the next entry.
+	Err() error
+
+	// Close should be called at the end of processing entries, required to release resources used to scan entries.
+	// After calling 'Close' the instance should not be used as the behaviour will not be defined.
+	Close()
+}
+
+// Entry holds a pair of key/value
+type Entry struct {
+	Key   []byte
+	Value []byte
+}
+
+func (e *Entry) String() string {
+	if e == nil {
+		return "Entry{nil}"
+	}
+	return fmt.Sprintf("Entry{%s, %s}", e.Key, e.Value)
+}
+
+// map drivers implementation
+var (
+	drivers   = make(map[string]Driver)
+	driversMu sync.RWMutex
+)
+
+func Register(name string, driver Driver) {
+	if name == "" {
+		panic("kv store register name is missing")
```

Comment:
Being able to panic probably means this func needs godoc that says as much.  Callers will not read the function, but just might read the godoc.